### ISSUE TITLE
/rollback and /rollback_check are horribly slow

### ISSUE THREAD

**Issue by iqualfragile** [luanti-org/luanti#1613]
horribly slow. on sqlite as backend, maybe its better with others.

**Comment by iqualfragile** [comment]
Ok, so i took a look at the rollback databases structure, and i have one idea, that should lead to a lot of improvement: currently minetest uses

    CREATE INDEX IF NOT EXISTS `actionIndex` ON `action`(`x`,`y`,`z`,`timestamp`,`actor`)

which is horribly horrible for sqlite, as the documentation clearly states that indices on colums with not quite unique values (such as x,y and z) are inefficient, and only improve a bit by running analyze (which is not done by default), indexing spatial information with b-trees is a bad idea anyway.
also combined indices… well might not be the best thing either.
The right way to go about this would be to use an [r* tree](https://www.sqlite.org/rtree.html) for the coordinates and two normal b-tree indices for the timestamp and actor.

**Comment by ShadowNinja** [comment]
@iqualfragile: I made a quick and dirty implementation of the R*Tree here: https://github.com/ShadowNinja/minetest/tree/rollback-rtree
It's hacky and has no backwards compatability, but it should work.
Do you have a database that you could convert and test?
Conversion:
  * Make the `actionTimeActor` index
  * Drop any other indexes (not strictly necessary).
  * Add the R*Tree virtual table.
  * Something like this:

```SQL
CREATE TABLE action_tmp (
	`id` INTEGER PRIMARY KEY AUTOINCREMENT,
	`actor` INTEGER NOT NULL,
	`timestamp` TIMESTAMP NOT NULL,
	`type` INTEGER NOT NULL,
	`list` TEXT,
	`index` INTEGER,
	`add` INTEGER,
	`stackNode` INTEGER,
	`stackQuantity` INTEGER,
	`nodeMeta` INTEGER,
	`oldNode` INTEGER,
	`oldParam1` INTEGER,
	`oldParam2` INTEGER,
	`oldMeta` TEXT,
	`newNode` INTEGER,
	`newParam1` INTEGER,
	`newParam2` INTEGER,
	`newMeta` TEXT,
	`guessedActor` INTEGER,
	FOREIGN KEY (`actor`) REFERENCES `actor`(`id`),
	FOREIGN KEY (`stackNode`) REFERENCES `node`(`id`),
	FOREIGN KEY (`oldNode`)   REFERENCES `node`(`id`),
	FOREIGN KEY (`newNode`)   REFERENCES `node`(`id`)
);
-- This might take a while
INSERT INTO `action_tmp` (<action_tmp field names>) SELECT (<action_tmp field names>) FROM action;
DROP TABLE `action`;
ALTER TABLE `action_tmp` RENAME TO `action`;
-- This too even
INSERT INTO `actionPos` (`id`, `x1`, `x2`, `y1`, `y2`, `z1`, `z2`) SELECT (`id`, `x`, `x`, `y`, `y`, `z`, `z`) FROM `action` WHERE `x` IS NOT NULL;
```

**Comment by iqualfragile** [comment]
oh my god
this is so much better
like a range-1000 search takes <5 seconds where before it took multiple minutes

**Comment by est31** [comment]
I think that 380e1504eb2445fd43c03c87d23b8e41e85a7871 has resolved this issue.
The comment above seems to indicate that the issue creator sees it solved as well.

**Comment by est31** [comment]
Ah sorry, it seems the comment only was about @ShadowNinja 's changes. Reopening.

Want to make it backwards compatible & do a PR?

**Comment by nerzhul** [comment]
i don't think we can do something on this, it's just heavy in terms of database changes

**Comment by sfan5** [comment]
fixed by 7c6ade0 according to @cheapie
